test/jdk/com/sun/jndi/dns/ConfigTests/Timeout.java

```diff
@@ -92,28 +95,32 @@ public void runTest() throws Exception {
     @Override
     public boolean handleException(Exception e) {
         if (e instanceof CommunicationException) {
-            Duration elapsedTime = Duration.between(startTime, Instant.now());
+            Duration elapsedTime = Duration.ofNanos(System.nanoTime() - startTime);
             if (!(((CommunicationException) e)
                     .getRootCause() instanceof SocketTimeoutException)) {
                 return false;
             }
 
             Duration expectedTime = Duration.ofMillis(TIMEOUT)
-                    .multipliedBy((1 << RETRIES) - 1);
+                    .multipliedBy((1 << RETRIES) - 1)
+                    .minus(Duration.ofMillis((DNS_CLIENT_MIN_TIMEOUT - 1) * RETRIES));
             DNSTestUtils.debug("Elapsed (ms):  " + elapsedTime.toMillis());
             DNSTestUtils.debug("Expected (ms): " + expectedTime.toMillis());
 
             // Check that elapsed time is as long as expected, and
-            // not more than 50% greater.
+            // not more than 67% greater. Given the min DNS timeout
+            // correction above the threshold value is equal to 61%.
```

Comment:
I don't think it is a rabbit hole, to provide some additional clarity on the timeout mechanism and to avoid any perception that it is an absolute realtime timeout semantics, such that developers have precise view of how the mechanism works

https://docs.oracle.com/en/java/javase/22/docs/api/jdk.naming.dns/module-summary.html

Reply:
I think 2 times is good, remove all potential noise  ;-)

the following failures is nearly twice the expected

----------System.out:(3/73)----------
Skip local DNS Server creation
Elapsed (ms): 14229
Expected (ms): 7750
----------System.err:(13/652)----------
java.lang.RuntimeException: Failed: timeout in 14229 ms, expected 7750ms
at Timeout.handleException(Timeout.java:116)
at TestBase.launch(TestBase.java:84)
at TestBase.run(TestBase.java:50)
at Timeout.main(Timeout.java:59)

and  I think it fails  the new upper bound check 

most of the elapsed times that have been less than the expected have been within the 50 * 5  tolerance, but there have been a few outside the -250 mess lower bound

Reply:
I agree that we don't want to document too much here. Updated the factor to 1.75 (2 seems a bit high and might hide real issues), and to make the timeout value calculation and check less arcane - I have updated test output to print the range of acceptable timeout values: 05ed9e053865293a1938ed7bc6fe208759513813

Reply:
2 time  is not too high,
I have presented, in the comment, a failures with the elapsed time is almost twice the expected time
where the elapsed time is 14229 !! which is approx 1.84 * expected timeout

Reply:
@msheppar with the latest proposed change to set MIN_TIMEOUT to 0, then I believe the `@implNote` you suggested is no longer needed.

Reply:
To address issue `#3` the UDP part of the `DnsClient` implementation has been changed to track the timeout left for each server for the unfulfilled timeout left for each server. An additional retry attempt was added to compensate for left timeouts.
The exception handling has been also updated to save observed exceptions as suppressed.
Also, the TCP part of the `DnsClient` has been changed to use nano time source for timeout calculations.